Core/GameEngine/Source/Common/System/MiniDumper.cpp

```diff
@@ -0,0 +1,710 @@
+		return;
+	}
+
+	PMINIDUMP_EXCEPTION_INFORMATION exceptionInfoPtr = NULL;
+	MINIDUMP_EXCEPTION_INFORMATION exceptionInfo = { 0 };
+	if (g_dumpException != NULL)
+	{
+		exceptionInfo.ExceptionPointers = g_dumpException;
+		exceptionInfo.ThreadId = g_dumpExceptionThreadId;
+		exceptionInfo.ClientPointers = FALSE;
+		exceptionInfoPtr = &exceptionInfo;
+	}
+
+	PMINIDUMP_CALLBACK_INFORMATION callbackInfoPtr = NULL;
+	MINIDUMP_CALLBACK_INFORMATION callBackInfo = { 0 };
+	if (dumpType == DUMP_TYPE_GAMEMEMORY)
+	{
+		callBackInfo.CallbackRoutine = MiniDumpCallback;
+		callBackInfo.CallbackParam = this;
+		callbackInfoPtr = &callBackInfo;
+	}
+
+	int dumpTypeFlags = MiniDumpNormal;
+	switch (dumpType)
+	{
+	case DUMP_TYPE_FULL:
+		dumpTypeFlags |= MiniDumpWithFullMemory;
+		FALLTHROUGH;
+	case DUMP_TYPE_GAMEMEMORY:
+		dumpTypeFlags |= MiniDumpWithDataSegs | MiniDumpWithHandleData | MiniDumpWithThreadInfo | MiniDumpWithFullMemoryInfo | MiniDumpWithPrivateReadWriteMemory;
+		FALLTHROUGH;
+	case DUMP_TYPE_MINIMAL:
+		dumpTypeFlags |= MiniDumpWithIndirectlyReferencedMemory | MiniDumpScanMemory;
+		break;
+	}
+
+	MINIDUMP_TYPE miniDumpType = static_cast<MINIDUMP_TYPE>(dumpTypeFlags);
+	BOOL success = DbgHelpLoader::miniDumpWriteDump(
+		::GetCurrentProcess(),
+		currentProcessId,
+		dumpFile,
+		miniDumpType,
+		exceptionInfoPtr,
+		NULL,
+		callbackInfoPtr);
+
+	if (!success)
+	{
+		DEBUG_LOG(("MiniDumper::CreateMiniDump: Unable to write minidump file '%s': error=%u", m_dumpFile, ::GetLastError()));
+	}
+	else
+	{
+		DEBUG_LOG(("MiniDumper::CreateMiniDump: Successfully wrote minidump file to '%s'", m_dumpFile));
+	}
+
+	::CloseHandle(dumpFile);
+}
+
+BOOL CALLBACK MiniDumper::MiniDumpCallback(PVOID CallbackParam, PMINIDUMP_CALLBACK_INPUT CallbackInput, PMINIDUMP_CALLBACK_OUTPUT CallbackOutput)
+{
+	if (CallbackParam == NULL || CallbackInput == NULL || CallbackOutput == NULL)
+	{
+		DEBUG_LOG(("MiniDumper::MiniDumpCallback: Required parameters were null; CallbackParam=%p, CallbackInput=%p, CallbackOutput=%p.",
+			CallbackParam, CallbackInput, CallbackOutput));
+		return false;
+	}
+
+	MiniDumper* dumper = static_cast<MiniDumper*>(CallbackParam);
+	return dumper->CallbackInternal(*CallbackInput, *CallbackOutput);
+}
+
+Bool MiniDumper::ShouldWriteDataSegsForModule(const PWCHAR module) const
+{
+	// Only include data segments for the game, ntdll and kernel32 modules to keep dump size low
+	static constexpr const WideChar* wanted_modules[] = { L"ntdll.dll", L"kernel32.dll"};
+	if (endsWithNoCase(module, m_executablePath))
+	{
+		return true;
+	}
+
+	for (size_t i = 0; i < ARRAY_SIZE(wanted_modules); ++i)
+	{
+		if (endsWithNoCase(module, wanted_modules[i]))
+		{
+			return true;
+		}
+	}
+
+	return false;
+}
+
+// This is where the memory regions and things are being filtered
+BOOL MiniDumper::CallbackInternal(const MINIDUMP_CALLBACK_INPUT& input, MINIDUMP_CALLBACK_OUTPUT& output)
+{
+	BOOL retVal = TRUE;
```

Comment:
What does the `retVal` signal to the caller? Perhaps give it a name that describes it, because it is not obvious what return TRUE and FALSE will do now.